text/0000-vec-append-from-within.md

```diff
@@ -0,0 +1,103 @@
+- Feature Name: `vec_append_from_within`
+- Start Date: 2019-06-21
+- RFC PR: [rust-lang/rfcs#2714](https://github.com/rust-lang/rfcs/pull/2714)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Provide a safe and efficient way append some of elements of a vector to itself. This is similar to the recently stabilized [`slice::copy_within()`](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_within), but appends to the vector instead of copying to a location within it.
+
+# Motivation
+[motivation]: #motivation
+
+Copying parts of a data stream to its end is an essential operation in decompressors and multimedia decoders - an area that could hugely benefit from Rust's performance and safety guarantees. Even though it requires just a little unsafe code, in practice people are struggling to implement it correctly. Motivating examples:
+
+ * Relevant code in `inflate` crate was vulnerable (memory disclosure). See [announcement](https://www.reddit.com/r/rust/comments/8zpp5f/) and a [simplified example](https://github.com/rust-lang/rfcs/pull/2714#issuecomment-504671231).
+ * A vulnerability in such code in `libflate` is currently pending disclosure, see https://github.com/sile/libflate/issues/33.
+ * **Rust standard library itself** had a buffer overflow bug in exactly this code. This is known as CVE-2018-1000810. See [announcement](https://blog.rust-lang.org/2018/09/21/Security-advisory-for-std.html) and the [fix](https://github.com/rust-lang/rust/pull/54397).
+
+This proposal is an attempt to provide the minimum viable building block that will allow safe and efficient implementations of RLE and similar predictive algorithms.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+```rust
+assert_eq!(vec![3,5,7].append_from_within((..1)), vec![3,5,7,3]);
+assert_eq!(vec![3,5,7].append_from_within((1..)), vec![3,5,7,5,7]);
+assert_eq!(vec![3,5,7].append_from_within((..)), vec![3,5,7,3,5,7]);
```

Comment:
nits: the extra parens around the argument confused me, and the method doesn't return anything (based on the signature below) so these `assert_eq!`s wouldn't compile.

Reply:
:man_facepalming: 
Fixed, thanks!